### ISSUE TITLE
Do's and Don'ts for contributors

### ISSUE THREAD

**Issue by corneliusweig** [GoogleContainerTools/skaffold#1945]
### tl;dr;
It would be helpful to have some more guidelines how to interact with the core team.

### Unanswered questions
After several PRs for the Skaffold project, I realized that I still don't know the best practices how to interact with the core team. So I compiled a list of questions:

- Should I generally use @-mentioning sparingly to avoid constant noise, or is it fine?
- After addressing review feedback, should I @-mention the reviewer?
- Would it help if I request review from a specific person? For example, when submitting a PR for an issue where Mrs.CoreX has commented, should I @-mention her?
- How can I get rid of e.g. `needs-rebase` label? Should I @-mention the person who has added the label?
- Should I always open an issue to discuss and plan a change, or is it ok to submit just a PR?
- What git workflow is expected for my PRs with merge conflicts? Should I rebase or merge master?
- When adding further changes to a PR, should I make reviewer-friendly commits (add a new one) or amend my PR commits history?
  - if the answer is "add new ones": is it encouraged to tidy up the git history by interactive rebase when the reviews are done?
- For docs PRs: how do I add the docs-modification label?
- When I realize my PR is insufficient, should I add _WIP_ in the title?
- If my PR is idle for &lt;n&gt; days, can I make it more visible?
- Should I generally not work on issues that have somebody assigned (even if that assignment is quite old).

Ultimately, my goal is to improve the collaboration of the community with the core team, and of course to reduce the time that community-PRs are open. 

### What would help me
I do realize, that many of the above questions may not have universal answers. But those that do should be documented in the CONTRIBUTING document in Skaffold.

Also, the core-team members will probably have a similar list of pain points where community guidelines could help. Please be blunt in what causes you friction, otherwise we can't know.

**Comment by balopat** [comment]
Thank you for filing this @corneliusweig - first of all, it is great that you took the courage to show your side of the experience as our contributor and start the conversation around it. We do care about having a great community and to be frank, we are still learning how to juggle managing the community with developing features. There is a constantly increasing amount of information that we have to filter through and things do fall through the cracks. This is not going to get better on the long run - as the number of users grow faster than our team.

I'll answer some inline that I can, the rest will be subject of discussions: 

> * Should I generally use @-mentioning sparingly to avoid constant noise, or is it fine?

It is fine. 

> * After addressing review feedback, should I @-mention the reviewer?

Sure, it can help, though for me personally, I'm really bad at notifications - so it might not help. For some contributors it might, it is okay to do it, and it will increase your chances. 

> * Would it help if I request review from a specific person? For example, when submitting a PR for an issue where Mrs.CoreX has commented, should I @-mention her?

Again, it can help, and doesn't hurt anything. It can't guarantee anything either though. 

> * How can I get rid of e.g. `needs-rebase` label? Should I @-mention the person who has added the label?

Unfortunately to manage labels you have to be a contributor on the repo (btw I think you specifically _might_ have permission?) - which is not going to change unless we switch over to something like Prow that would help adding/removing labels/statuses through a separate set of commands.

> * For docs PRs: how do I add the docs-modification label?

See above for labels. However we are planning to move to a more automated mode for previews. 

> * Should I always open an issue to discuss and plan a change, or is it ok to submit just a PR?

Rule of thumb: an issue is always best to have first. But it depends on the size, if it's a tiny PR, you should be okay without an issue. This is something we should capture more clearly in the contrib guide, I agree. 

> * What git workflow is expected for my PRs with merge conflicts? Should I rebase or merge master?

I think merge is better 
  - we do checkout code from forks - updating is a bit more painful when you rebase
  - we don't really care about the commit history on the PR - all our merges are squashed anyway

> * When adding further changes to a PR, should I make reviewer-friendly commits (add a new one) or amend my PR commits history?

See above - I'd prefer commits, especially after a couple of round of reviews to avoid painful checkouts. But initially amends should be fine. 

>   * if the answer is "add new ones": is it encouraged to tidy up the git history by interactive rebase when the reviews are done?

I really rarely use the commit history - others can chime in here. 

> * When I realize my PR is insufficient, should I add _WIP_ in the title?

Yes, that would be a helpful signal - that will stop the reviews and we can focus on something else until you sort out the WIP. 

> * If my PR is idle for <n> days, can I make it more visible?

Not sure if you can do anything about this besides pinging via mentions - we are trying to get better at PR times.

> * Should I generally not work on issues that have somebody assigned (even if that assignment is quite old).

Ask first the assignee, if no answer in a week (maybe holidays?) then probably safe to pickup - we did have a bunch of abandonded but assigned issues. 

> Ultimately, my goal is to improve the collaboration of the community with the core team, and of course to reduce the time that community-PRs are open.

Again, thank you!

**Comment by corneliusweig** [comment]
First, thanks a lot for your clarifications, I hope this will make it into some official contributing document in the repo!

Just to answer the question about labels: despite being a collaborator, I cannot modify labels.

About squash merge: I think this is handled differently in the core team :smiley: . Me personally, I find it a pity if commits are always squashed. For small PRs it sure is fine (especially when it is about a few fixups). But in general I think this is a lost opportunity to add more meaning to the code. For example, when I want to know why a particular line is the way it is, I always annotate the code and check the commit message. This is most helpful if the respective commit was a small logical unit.

**Comment by tstromberg** [comment]
Marking as stale, since it's been a while and does not appear to be on the teams priority radar. Feel free to reopen if you disagree.

I personally feel like this should get captured into a `CONTRIBUTORS.md` doc.
